### PR TITLE
Proposal: `pre-commit` for sigopt-server

### PR DESCRIPTION
**Pull request by tjs-intel**
`pre-commit` can handle, as the name suggests, `pre-commit` tasks, but also pre-push tasks like making sure we don't commit or push to the `main` branch. The `pre-commit-hooks` repo has a few convenient hooks, including a few that replace rules that we have custom implementations for:

```
      - id: check-added-large-files
      - id: check-ast
      - id: check-executables-have-shebangs
      - id: check-merge-conflict # find merge conflict strings in files
      - id: check-shebang-scripts-are-executable
      - id: check-symlinks # makes sure symlinks are valid
      - id: check-toml
      - id: check-vcs-permalinks # makes sure vcs links are permalinks
      - id: check-yaml
      - id: debug-statements # look for import pdb etc.
      - id: destroyed-symlinks # look for files
      - id: detect-private-key
      - id: end-of-file-fixer # makes sure files end in newlines
      - id: mixed-line-ending # makes sure files have consistent end-of-line sequences
      - id: no-commit-to-branch # dont commit to main branch
      - id: requirements-txt-fixer # sorts requirements*.txt files
      - id: trailing-whitespace # trims trailing whitespace from files
```

There are also integrations for the various linting tools that we use, namely:
 * bandit (native)
 * black (native)
 * flake8 (native)
 * hadolint (native)
 * isort (native)
 * mypy (mirror)
 * prettier (mirror)
 * vulture (native)
 * eslint (native support for pre-commit, but better for us to run it natively as to not duplicate package.json)
 * pylint (same support as eslint)

In addition I added the following tools to `pre-commit`:
 * circleci-config-validate: validates circleci configuration file structure
 * apply-copyright-notice: apply the copyright and license notice
 * custom-bash-lint: check for `set -e` and `set -o pipefail`
 * custom-python-lint: apply out custom python lint rules

`pre-commit` also has a CI tool, pre-commit.ci, which is currently active on `libsigopt`. It runs `pre-commit` on each PR and patches the PR when there are any changes made by the hooks.

We currently use `lint-staged`, which has served us well, but does not have as good multi-language support. `pre-commit` hooks are configured to apply to file `types`, which match to tags produced by the `identify` library https://pypi.org/project/identify/, `lint-staged` only supports regex matching of file names.

A discrepancy from `lint-staged`: attempting to commit badly formatted code with `pre-commit` will cause the commit to fail. The files will be well formatted though and the developer just needs to check them then re-attempt the commit.

Developers can opt-out of `pre-commit` by simply not running `make setup-hooks`. In this case it is possible to run `pre-commit run` to run all the same checks on modified files.

### CODE REVIEW

File: src/python/zigopt/utils/create_produser.py [hunk 1]
```diff
@@ -90,63 +89,3 @@ def execute_query(conn, cursor, query_string, query_args=None, error_allow_list=
...
-  parser.add_argument(
-    "--produser_password",
-    default="produser_password_for_development",
-    help="The produser's password",
-  )
-
-  return parser.parse_args()
-
-
-if __name__ == "__main__":
```
Comment (simonh-intel) [inline src/python/zigopt/utils/create_produser.py:142]: This change seems unrelated and or odd?

**Review by simonh-intel — APPROVED** [review]
[create_produser.py](https://github.com/sigopt/sigopt-server/pull/82/files#diff-131acc6a47b260ce10ac3a6b66be8d67dff85d1abf2b6f1de9cf969aad159ef2)

change seems a bit odd to me but otherwise LGTM

File: src/python/zigopt/utils/create_produser.py (hunk 1 above)
Comment (tjs-intel) [reply src/python/zigopt/utils/create_produser.py:142]: This script was broken because `query` was a missing required argument for `create_produser`. This script is unused anyway

File: .pylintrc
Comment (dwanderson-intel) [inline .pylintrc:1]: any chance we can move this back to a visible file under tools? I feel that having explicit, and non-hidden config files makes it easier to track down, rather than having to divine that we have such a file hanging out somewhere

Comment (tjs-intel) [reply .pylintrc:1]: I'd argue that it's even more hidden by putting it in a non-standard place. I'll look into putting it in `pyproject.toml` or `setup.cfg`

File: setup.cfg
Comment (dwanderson-intel) [inline setup.cfg:1]: oof, definitely would like this file renamed: `setup.cfg` isn't clearly connected to `isort`

Comment (dwanderson-intel) [reply setup.cfg:1]: oh but it also has `flake8`? i'll re-read the PR message to see if this is more clear

Comment (tjs-intel) [reply setup.cfg:1]: `setup.cfg` seems to be a standard location for configuring tools. `pyproject.toml` is also standard, but it seems that not all projects support just `setup.cfg` or `pyproject.toml`

File: .pylintrc
Comment (dwanderson-intel) [reply .pylintrc:1]: Yeah, I can see it both ways - I think when we just had a plethora of dotfiles handling many different linters and etc it got a bit overwhelming; doesn't look like we have too many anymore, so I'm more ok with this.

File: setup.cfg
Comment (dwanderson-intel) [reply setup.cfg:1]: got it, im good with this now!
